File: Tree/568-minDiffInBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        ans=inf
        prev=None
        def dfs(root):
            if not root:
                return 
            dfs(root.left)
            nonlocal prev,ans
            if prev:
                ans=min(ans,root.val-prev.val)
            prev=root
            dfs(root.right)
        dfs(root)
        return ans


        # nums=[]
        # ans=inf
        # def inorder(root):
        #     if not root:
        #         return 

        #     inorder(root.left)
        #     nums.append(root.val)
        #     inorder(root.right)
        
        # inorder(root)
        # for i in range(1,len(nums)):
        #     ans=min(ans,nums[i]-nums[i-1])
        # return ans

File: Tree/568-minDiffInBST.cpp
#include<iostream>
#include<cmath>
class TN
{
    public:
        int m_v{};
        TN* left{};
        TN* right{};
        TN(int v,TN* l=nullptr, TN* r=nullptr):m_v{v},left{l},right{r}{};
};

void dfoo(const TN* root,const TN* prev,int &ans)
{
    if(!root)
        return ;
    dfoo(root->left,prev,ans);
    if(prev)
    {
        ans=std::min(ans,root->m_v-prev->m_v);
    }
    prev=root;
    dfoo(root->right,prev,ans);
}

int minDiff(const TN* root)
{
    TN* prev=nullptr;
    int ans=INT_MAX;
    dfoo(root,prev,ans);
    return ans;
}

int main()
{
    TN *root=new TN(10);
    TN *one=new TN(1);
    TN *two=new TN(12);
 

    root->left=one;
    // two->right=one;
    root->right=two;
    std::cout<<minDiff(root);
}
